Got out-of-order AST nodes(s) with (python typing + async def)

Hi,

Thanks for the awesome project. 

```bash
syntax$ tidy-imports --version
pyflyby 1.7.2 (tidy-imports)
syntax$ cat a.py
import aiofiles


class A():
    async def func(self, location: str) -> bytes:
        async with aiofiles.open(location, "rb") as file:
            return await file.read()
```

When I tried to tydi-imports for some code above, I got error below

```bash
syntax$ tidy-imports a.py
Traceback (most recent call last):
  File "/home/ys/.asdf/installs/python/3.9.6/lib/python3.9/site-packages/pyflyby/_cmdline.py", line 384, in process_actions
    action(m)
  File "/home/ys/.asdf/installs/python/3.9.6/lib/python3.9/site-packages/pyflyby/_cmdline.py", line 422, in action_ifchanged
    if m.output_content.joined == m.input_content.joined:
  File "/home/ys/.asdf/installs/python/3.9.6/lib/python3.9/site-packages/pyflyby/_util.py", line 62, in __get__
    result = self.method(inst)
  File "/home/ys/.asdf/installs/python/3.9.6/lib/python3.9/site-packages/pyflyby/_cmdline.py", line 334, in output_content
    return FileText(self.modifier(self.input_content), filename=self.filename)
  File "/home/ys/.asdf/installs/python/3.9.6/bin/tidy-imports", line 121, in modify
    x = canonicalize_imports(x, params=options.params)
  File "/home/ys/.asdf/installs/python/3.9.6/lib/python3.9/site-packages/pyflyby/_imports2s.py", line 606, in canonicalize_imports
    return transform_imports(codeblock, transformations, params=params)
  File "/home/ys/.asdf/installs/python/3.9.6/lib/python3.9/site-packages/pyflyby/_imports2s.py", line 565, in transform_imports
    transformer = SourceToSourceFileImportsTransformation(codeblock)
  File "/home/ys/.asdf/installs/python/3.9.6/lib/python3.9/site-packages/pyflyby/_imports2s.py", line 26, in __new__
    return cls._from_source_code(arg)
  File "/home/ys/.asdf/installs/python/3.9.6/lib/python3.9/site-packages/pyflyby/_imports2s.py", line 34, in _from_source_code
    self.preprocess()
  File "/home/ys/.asdf/installs/python/3.9.6/lib/python3.9/site-packages/pyflyby/_imports2s.py", line 90, in preprocess
    for is_imports, subblock in self.input.groupby(lambda ps: ps.is_import):
  File "/home/ys/.asdf/installs/python/3.9.6/lib/python3.9/site-packages/pyflyby/_parse.py", line 1305, in groupby
    for pred, stmts in groupby(self.statements, predicate):
  File "/home/ys/.asdf/installs/python/3.9.6/lib/python3.9/site-packages/pyflyby/_util.py", line 62, in __get__
    result = self.method(inst)
  File "/home/ys/.asdf/installs/python/3.9.6/lib/python3.9/site-packages/pyflyby/_parse.py", line 1245, in statements
    node = self.annotated_ast_node
  File "/home/ys/.asdf/installs/python/3.9.6/lib/python3.9/site-packages/pyflyby/_util.py", line 62, in __get__
    result = self.method(inst)
  File "/home/ys/.asdf/installs/python/3.9.6/lib/python3.9/site-packages/pyflyby/_parse.py", line 1168, in annotated_ast_node
    _annotate_ast_nodes(result)
  File "/home/ys/.asdf/installs/python/3.9.6/lib/python3.9/site-packages/pyflyby/_parse.py", line 295, in _annotate_ast_nodes
    _annotate_ast_startpos(ast_node, None, startpos, text, flags)
  File "/home/ys/.asdf/installs/python/3.9.6/lib/python3.9/site-packages/pyflyby/_parse.py", line 370, in _annotate_ast_startpos
    leftstr = _annotate_ast_startpos(child_node, ast_node,
  File "/home/ys/.asdf/installs/python/3.9.6/lib/python3.9/site-packages/pyflyby/_parse.py", line 370, in _annotate_ast_startpos
    leftstr = _annotate_ast_startpos(child_node, ast_node,
  File "/home/ys/.asdf/installs/python/3.9.6/lib/python3.9/site-packages/pyflyby/_parse.py", line 376, in _annotate_ast_startpos
    raise AssertionError(
AssertionError: While processing /home/ys/Practice/languages/python/syntax/a.py: Got out-of-order AST node(s):
  parent minpos=(1,1)
    node: AsyncFunctionDef(name='func', args=arguments(posonlyargs=[], args=[arg(arg='self'), arg(arg='location', annotation=Name(id='str', ctx=Load()))], kwonlyargs=[], kw_defaults=[], defaults=[]), body=[AsyncWith(items=[withitem(context_expr=Call(func=Attribute(value=Name(id='aiofiles', ctx=Load()), attr='open', ctx=Load()), args=[Name(id='location', ctx=Load()), Constant(value='rb')], keywords=[]), optional_vars=Name(id='file', ctx=Store()))], body=[Return(value=Await(value=Call(func=Attribute(value=Name(id='file', ctx=Load()), attr='read', ctx=Load()), args=[], keywords=[])))])], decorator_list=[], returns=Name(id='bytes', ctx=Load()))
      fields: name args body decorator_list returns type_comment
      children:
                     : arguments(posonlyargs=[], args=[arg(arg='self'), arg(arg='location', annotation=Name(id='str', ctx=Load()))], kwonlyargs=[], kw_defaults=[], defaults=[])
                (6,9): AsyncWith(items=[withitem(context_expr=Call(func=Attribute(value=Name(id='aiofiles', ctx=Load()), attr='open', ctx=Load()), args=[Name(id='location', ctx=Load()), Constant(value='rb')], keywords=[]), optional_vars=Name(id='file', ctx=Store()))], body=[Return(value=Await(value=Call(func=Attribute(value=Name(id='file', ctx=Load()), attr='read', ctx=Load()), args=[], keywords=[])))])
        ==>    (5,44): Name(id='bytes', ctx=Load())

This indicates a bug in pyflyby._

pyflyby developer: Check if there's a bug or missing ast node handler in pyflyby._parse._iter_child_nodes_in_order() - probably the handler for ast.AsyncFunctionDef.

/home/ys/.asdf/installs/python/3.9.6/bin/tidy-imports: encountered the following problems:
    /home/ys/Practice/languages/python/syntax/a.py: AssertionError: Got out-of-order AST node(s):              parent minpos=(1,1)
                node: AsyncFunctionDef(name='func', args=arguments(posonlyargs=[], args=[arg(arg='self'), arg(arg='location', annotation=Name(id='str', ctx=Load()))], kwonlyargs=[], kw_defaults=[], defaults=[]), body=[AsyncWith(items=[withitem(context_expr=Call(func=Attribute(value=Name(id='aiofiles', ctx=Load()), attr='open', ctx=Load()), args=[Name(id='location', ctx=Load()), Constant(value='rb')], keywords=[]), optional_vars=Name(id='file', ctx=Store()))], body=[Return(value=Await(value=Call(func=Attribute(value=Name(id='file', ctx=Load()), attr='read', ctx=Load()), args=[], keywords=[])))])], decorator_list=[], returns=Name(id='bytes', ctx=Load()))
                  fields: name args body decorator_list returns type_comment
                  children:
                                 : arguments(posonlyargs=[], args=[arg(arg='self'), arg(arg='location', annotation=Name(id='str', ctx=Load()))], kwonlyargs=[], kw_defaults=[], defaults=[])
                            (6,9): AsyncWith(items=[withitem(context_expr=Call(func=Attribute(value=Name(id='aiofiles', ctx=Load()), attr='open', ctx=Load()), args=[Name(id='location', ctx=Load()), Constant(value='rb')], keywords=[]), optional_vars=Name(id='file', ctx=Store()))], body=[Return(value=Await(value=Call(func=Attribute(value=Name(id='file', ctx=Load()), attr='read', ctx=Load()), args=[], keywords=[])))])
                    ==>    (5,44): Name(id='bytes', ctx=Load())

            This indicates a bug in pyflyby._

            pyflyby developer: Check if there's a bug or missing ast node handler in pyflyby._parse._iter_child_nodes_in_order() - probably the handler for ast.AsyncFunctionDef.
```

I bisected my codes to narrow problematic syntax and it seems to be "typing annotation". Could you please take a look if it's possible to parse the syntax?

I checked syntaxs below were working.
```
import aiofiles


class A(): 
    # Without type annotation
    async def func(self, location):
        async with aiofiles.open(location, "rb") as file:
            return await file.read()

```

```
import aiofiles


class A():
    # Without async in front of func()
    def func(self, location: str) -> bytes:
        async with aiofiles.open(location, "rb") as file:
            return await file.read()

```

Thanks!

--- a/lib/python/pyflyby/_parse.py
+++ b/lib/python/pyflyby/_parse.py
@@ -31,12 +31,17 @@ else:
 
 
 if sys.version_info >= (3, 8):
-    from ast import TypeIgnore
+    from ast import TypeIgnore, AsyncFunctionDef
 else:
-    # TypeIgnore does not exist on Python 3.7 and before.
-    # thus we define a dummy TypeIgnore just to simplify remaining code.
+
+    # TypeIgnore, AsyncFunctionDef does not exist on Python 3.7 and before. thus
+    # we define a dummy TypeIgnore, AsyncFunctionDef just to simplify remaining
+    # code.
 
     class TypeIgnore:
+        pass
+
+    class AsyncFunctionDef:
         pass
 
 
@@ -104,7 +109,7 @@ def _iter_child_nodes_in_order_internal_1(node):
     if isinstance(node, ast.Dict):
         assert node._fields == ("keys", "values")
         yield list(zip(node.keys, node.values))
-    elif isinstance(node, ast.FunctionDef):
+    elif isinstance(node, (ast.FunctionDef, AsyncFunctionDef)):
         if six.PY2:
             assert node._fields == ('name', 'args', 'body', 'decorator_list'), node._fields
             yield node.decorator_list, node.args, node.body
@@ -423,7 +428,7 @@ def _annotate_ast_startpos(ast_node, parent_ast_node, minpos, text, flags):
         # In Python 3.8+, FunctionDef.lineno is the line with the def. To
         # account for decorators, we need the lineno of the first decorator
         if (sys.version_info >= (3, 8)
-            and isinstance(ast_node, (ast.FunctionDef, ast.ClassDef))
+            and isinstance(ast_node, (ast.FunctionDef, ast.ClassDef, AsyncFunctionDef))
             and ast_node.decorator_list):
             delta = (ast_node.decorator_list[0].lineno-1,
                      # The col_offset doesn't include the @
@@ -755,7 +760,8 @@ def _ast_node_is_in_docstring_position(ast_node):
     it eligible as a docstring.
 
     The main way a ``Str`` can be a docstring is if it is a standalone string
-    at the beginning of a ``Module``, ``FunctionDef``, or ``ClassDef``.
+    at the beginning of a ``Module``, ``FunctionDef``, ``AsyncFucntionDef``
+    or ``ClassDef``.
 
     We also support variable docstrings per Epydoc:
 
@@ -783,7 +789,7 @@ def _ast_node_is_in_docstring_position(ast_node):
     if expr_ctx.field != 'body':
         return False
     parent_node = expr_ctx.parent
-    if not isinstance(parent_node, (ast.FunctionDef, ast.ClassDef, ast.Module)):
+    if not isinstance(parent_node, (ast.FunctionDef, ast.ClassDef, ast.Module, AsyncFunctionDef)):
         return False
     if expr_ctx.index == 0:
         return True
@@ -1369,7 +1375,7 @@ class PythonBlock(object):
         #    if _ast_node_is_in_docstring_position(n)]
         # However, the method we now use is more straightforward, and doesn't
         # require first annotating each node with context information.
-        docstring_containers = (ast.FunctionDef, ast.ClassDef, ast.Module)
+        docstring_containers = (ast.FunctionDef, ast.ClassDef, ast.Module, AsyncFunctionDef)
         for node in _walk_ast_nodes_in_order(self.annotated_ast_node):
             if not isinstance(node, docstring_containers):
                 continue

--- a/tests/test_parse.py
+++ b/tests/test_parse.py
@@ -557,17 +557,26 @@ def test_PythonBlock_flags_type_comment_fail_transform():
 
 
 
-def test_PythonBlock_flags_type_comment_ignore_fails_transform():
+examples_transform = ["""
+    a = None # type: ignore
+    """,
+    """
+    class A:
+        async def func(self, location: str) -> bytes:
+            async with aiofiles.open(location, "rb") as file:
+                return await file.read()
+        """
+]
+
+@pytest.mark.parametrize("source", examples_transform)
+def test_PythonBlock_flags_type_comment_ignore_fails_transform(source):
     """
     See https://github.com/deshaw/pyflyby/issues/174
 
     Type: ignore are custom ast.AST who have no col_offset.
     """
     block = PythonBlock(
-    dedent("""
-        a = None # type: ignore
-        """
-    ))
+    dedent(source))
     s = SourceToSourceFileImportsTransformation(block)
     assert s.output() == block
 

--- a/tests/test_parse.py
+++ b/tests/test_parse.py
@@ -559,14 +559,17 @@ def test_PythonBlock_flags_type_comment_fail_transform():
 
 examples_transform = ["""
     a = None # type: ignore
-    """,
+    """]
+
+if PY3:
+    examples_transform.append(
     """
     class A:
         async def func(self, location: str) -> bytes:
             async with aiofiles.open(location, "rb") as file:
                 return await file.read()
         """
-]
+        )
 
 @pytest.mark.parametrize("source", examples_transform)
 def test_PythonBlock_flags_type_comment_ignore_fails_transform(source):

